File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
def solution(edges):
    answer = [0, 0, 0, 0]
    
    nodes = {} # 정점 번호: [정점에서 나간 간선 수, 정점으로 들어온 간선 수]
    
    for edge in edges:
        a, b = edge
        # 나가는 간선 
        if not nodes.get(a):
            nodes[a] = [1, 0]
        else:
            nodes[a][0] += 1
        # 들어온 간선
        if not nodes.get(b):
            nodes[b] = [0, 1]
        else:
            nodes[b][1] += 1
    
    for key, values in nodes.items():
        outside, inside = values
        
        if outside >= 2 and inside == 0: # 생성한 정점
            answer[0] = key
        elif outside == 0 and inside > 0: # 막대 그래프
            answer[2] += 1
        elif outside >= 2 and inside >= 2: # 8자 그래프
            answer[3] += 1
        
    answer[1] = nodes[answer[0]][0] - answer[2] - answer[3]
    
    return answer

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
import java.util.*;

class Solution {
    public int[] solution(int[][] edges) {
        int[] answer = new int[4];
        
        // 정점 번호 : [나가는 간선 수, 들어오는 간선 수]
        Map<Integer, int[]> nodes = new HashMap<>();
        
        for (int[] edge : edges) {
            int a = edge[0], b = edge[1];
            
            // 초기값 설정
            nodes.putIfAbsent(a, new int[]{0, 0});
            nodes.putIfAbsent(b, new int[]{0, 0});
            
            nodes.get(a)[0]++; // a에서 나가는 간선 증가
            nodes.get(b)[1]++; // b로 들어오는 간선 증가
        }
        
        for (Map.Entry<Integer, int[]> entry: nodes.entrySet()) {
            int key = entry.getKey();
            int outCount = entry.getValue()[0];
            int inCount = entry.getValue()[1];
            
            if (outCount >= 2 && inCount == 0) {
                answer[0] = key;
            } else if (outCount == 0 && inCount > 0) { // 막대그래프
                answer[2]++;
            } else if (outCount >= 2 && inCount >= 2) {
                answer[3]++;
            }
        }
        
        answer[1] = nodes.get(answer[0])[0] - answer[2] - answer[3]; // 도넛 그래프 개수 계산
            
        return answer;
    }
}